ts/session/medium_group/index.ts

```diff
@@ -39,56 +59,844 @@ async function createSenderKeysForMembers(
+
+  const allDevices = _.flatten(
+    await Promise.all(
+      existingMembers.map(m => MultiDeviceProtocol.getAllDevices(m))
+    )
+  );
+
+  let existingKeys: Array<RatchetState> = [];
+
+  if (leaving.length > 0) {
+    // If we have leaving members, we have to re-generate ratchet
+    // keys for existing members
+    const otherKeys = await Promise.all(
+      allDevices.map(async device => {
+        return createSenderKeyForGroup(groupId, PubKey.cast(device));
+      })
+    );
+
+    newKeys = _.union(newKeys, otherKeys);
+  } else {
+    // We can reuse existing keys
+    existingKeys = await getExistingSenderKeysForGroup(groupId, allDevices);
+  }
+
+  return { existingKeys, newKeys };
+}
+
+async function getGroupSecretKey(groupId: string): Promise<Uint8Array> {
+  const groupIdentity = await window.Signal.Data.getIdentityKeyById(groupId);
+  return new Uint8Array(fromHex(groupIdentity.secretKey));
+}
+
+async function syncMediumGroup(group: ConversationModel) {
+  const ourPrimary = await UserUtil.getPrimary();
+
+  const groupId = group.get('id');
+  const members = group.get('members');
+
+  const secretKey = await getGroupSecretKey(groupId);
+
+  const allDevices = _.flatten(
+    await Promise.all(members.map(m => MultiDeviceProtocol.getAllDevices(m)))
+  );
+
+  const secondaryKeys = await MultiDeviceProtocol.getSecondaryDevices(
+    ourPrimary
+  );
+
+  const existingKeys = await getExistingSenderKeysForGroup(groupId, allDevices);
+  const senderKeysForSecondary = await Promise.all(
+    secondaryKeys.map(key => createSenderKeyForGroup(groupId, key))
+  );
+  const senderKeysContainer: SenderKeysContainer = {
+    existingKeys,
+    newKeys: senderKeysForSecondary,
+  };
+
+  const groupUpdate: GroupInfo = {
+    id: group.get('id'),
+    name: group.get('name'),
+    members: group.get('members'),
+    is_medium_group: true,
+    admins: group.get('groupAdmins'),
+    senderKeysContainer,
+    secretKey,
+  };
+
+  // Note: we send this to our primary device which will in effect will send to
+  // our secondary devices, actually ignoring the primary
+  await sendGroupUpdateForMedium(
+    { joiningMembers: [ourPrimary.key] },
+    groupUpdate
+  );
+}
+
+// Secondary devices are not expected to already have the group, so
+// we send messages of type NEW
+export async function syncMediumGroups(groups: Array<ConversationModel>) {
+  await Promise.all(groups.map(syncMediumGroup));
+}
+
+export async function initiateGroupUpdate(
+  groupId: string,
+  groupName: string,
+  members: Array<string>,
+  avatar: any
+) {
+  const { ConversationController } = window;
+
+  const convo = await ConversationController.getOrCreateAndWait(
+    groupId,
+    'group'
+  );
+
+  const isMediumGroup = convo.isMediumGroup();
+
+  const groupDetails = {
+    id: groupId,
+    name: groupName,
+    members,
+    active: true,
+    expireTimer: convo.get('expireTimer'),
+    avatar,
+    is_medium_group: isMediumGroup,
```

Comment:
Didn't @vincentbavitz changed all those field to be matching (camel case everywhere?). Just thinking about conflicts

Reply:
As discussed in Discord.